src/main/java/tdl/client/queue/QueueBasedImplementationRunner.java

```diff
@@ -44,78 +219,193 @@ public Builder withSolutionFor(String methodName, UserImplementation userImpleme
 
             // Debt - we only need this to consume message from the server
             deployProcessingRules
                     .on("display_description")
                     .call(params -> "OK")
-                    .then(publish());
+                    .build();
 
+            logToConsole("        QueueBasedImplementationRunner.Builder createDeployProcessingRules [end]");
             return deployProcessingRules;
         }
     }
 
     public void run() {
+        logToConsole("        QueueBasedImplementationRunner run [start]");
         audit.logLine("Starting client");
-        try (RemoteBroker remoteBroker = new RemoteBroker(config.getHostname(), config.getPort(), config.getUniqueId(), config.getRequestTimeoutMillis())){
-            //Design: We use a while loop instead of an ActiveMQ MessageListener to process the messages in order
+        try {
             audit.logLine("Waiting for requests");
-            Optional<Request> request = remoteBroker.receive();
-            while (request.isPresent()) {
-                request = applyProcessingRules(request.get(), deployProcessingRules, remoteBroker);
-            }
+            logToConsole("     QueueBasedImplementationRunner receive [start]");
+            List<Message> messages;
+            boolean continueFetching;
+            do {
+                messages = client.receiveMessage(receiveRequestQueueMessage).getMessages();
+                continueFetching = messages.size() > 0;
+                if (continueFetching) {
+                    logToConsole("     QueueBasedImplementationRunner messages: " + messages);
+                    for (Message message : messages) {
+                        logToConsole("message: " + message);
+                        logToConsole("payload: " + message.getBody());
+
+                        try {
+                            logToConsole("        QueueBasedImplementationRunner applyProcessingRules [start]");
+                            JsonNode jsonNode = mapper.readValue(message.getBody(), JsonNode.class);
+                            String payload = jsonNode.get("payload").asText();
+                            JsonRpcRequest jsonRpcRequest = gson.fromJson(payload, JsonRpcRequest.class);
+
+                            Request request = new Request(message, jsonRpcRequest);
+                            audit.startLine();
+                            audit.log(request);
+                            requests.add(request);
+
+                            //Obtain response from user
+                            Response response = deployProcessingRules.getResponseFor(request);
+                            audit.log(response);
+
+                            //Act
+                            if (response instanceof FatalErrorResponse) {
+                                audit.endLine();
+                            } else {
+                                respondTo(request, with(response));
```

Comment:
Given the fact that we are consuming messages in batches, sending and deleting should also be done in batches.

The order is:
- receive a batch of messages (AWS `client.receiveMessage(receiveRequestQueueMessage)`)
- process their responses (`deployProcessingRules.getResponseFor(request);`)
- publish the responses (AWS `SendMessageBatchRequest`)
- delete the requests (AWS `DeleteMessageBatchRequest`)

Reply:
Receiving in batches already happens and is controlled by config - there is not `ReceiveMessageBatchRequest` class to be used.

The others I'll take on board and I think we had discussed it, can apply them now.

Reply:
My one comment on batch response sending, this will have to scenarios:
- A batch of successful responses, zero failed responses and a batch of messages to be deleted
- A batch of some successful responses, some failed responses and remaining batch of messages to be deleted

For the the second scenarios are we sure we dont have any issues? I will come back with more background when I have finished implementing it.

Reply:
You are right, we need to extend the"  Scenario: Should not publish after an exception when processing a message" to cover multiple messages.

The scenario should actually be called "Should stop processing after an exception is encountered"
```
  Scenario: Should not publish after an exception when processing a message
    Given I receive the following requests:
      | payload                                   |
      | {"method":"increment","params":[1],"id":"X1"}  |
      | {"method":"sum","params":[0,1],"id":"X2"} |
      | {"method":"increment","params":[2],"id":"X3"}  |
    When I go live with the following processing rules:
      | method       | call            |
      | sum          | throw exception |
    Then the client should not consume any request
    And the client should not publish any response
    And the client should display to console:
      | output                                                                                    |
      | id = X1, req = increment(1), resp = 2  |
      | id = X1, req = sum(0, 1), error = "user implementation raised exception", (NOT PUBLISHED) |
```

The breakdown is:
- Receive: [ X1, X2, X3 ]
- Process: [ X1, X2 ]
- Publish: [ X1 ]
- Delete: [ X1 ]